mutiny-core/src/federation.rs

```diff
@@ -0,0 +1,612 @@
+                    Ok(o) => {
+                        process_outcome(
+                            o,
+                            process_pay_state_ln,
+                            invoice.clone(),
+                            false,
+                            30_000,
+                            Arc::clone(&self.logger),
+                        )
+                        .await
+                    }
+                    Err(_) => invoice.clone().into(),
+                }
+            }
+        };
+
+        match inv.status {
+            HTLCStatus::Succeeded => Ok(inv),
+            HTLCStatus::Failed => Err(MutinyError::RoutingFailed),
+            HTLCStatus::Pending => Err(MutinyError::PaymentTimeout),
+            HTLCStatus::InFlight => Err(MutinyError::PaymentTimeout),
+        }
+    }
+
+    pub async fn get_mutiny_federation_identity(&self) -> FederationIdentity {
+        FederationIdentity {
+            uuid: self.uuid.clone(),
+            federation_id: self.fedimint_client.federation_id(),
+        }
+    }
+}
+
+// A federation private key will be derived from `m/1'/N'/X'`, where:
+// - N is the network type
+// - X is the index of a specific federation.
+//
+// Federation will derive further keys from there.
+fn create_federation_secret(
+    xprivkey: ExtendedPrivKey,
+    federation_index: &FederationIndex,
+    network: Network,
+) -> Result<DerivableSecret, MutinyError> {
+    let context = Secp256k1::new();
+    let xpriv = xprivkey.derive_priv(
+        &context,
+        &DerivationPath::from(vec![
+            ChildNumber::from_hardened_idx(1)?,
+            ChildNumber::from_hardened_idx(network_to_bip44_chain(network))?,
+            ChildNumber::from_hardened_idx(federation_index.child_index)?,
+        ]),
+    )?;
+    let secret =
+        DerivableSecret::new_root(&xpriv.private_key.secret_bytes(), FEDIMINT_CLIENT_NONCE);
+    Ok(secret)
+}
+
+fn network_to_bip44_chain(network: Network) -> u32 {
+    match network {
+        Network::Bitcoin => 0,
+        Network::Testnet => 1,
+        Network::Signet => 1,
+        Network::Regtest => 1,
+    }
+}
+
+async fn extract_invoice_from_entry(
+    logger: Arc<MutinyLogger>,
+    entry: &OperationLogEntry,
+    hash: &sha256::Hash,
+    operation_id: OperationId,
+    lightning_module: &LightningClientModule,
+) -> Option<MutinyInvoice> {
+    let lightning_meta: LightningOperationMeta = entry.meta();
+
+    match lightning_meta.variant {
+        LightningOperationMetaVariant::Pay(pay_meta) => {
+            if pay_meta.invoice.payment_hash() == hash {
+                match lightning_module.subscribe_ln_pay(operation_id).await {
+                    Ok(o) => Some(
+                        process_outcome(
+                            o,
+                            process_pay_state_ln,
+                            pay_meta.invoice,
+                            false,
+                            30,
```

Comment:
this 30 number is here again but not multiplied by 1k, probably a bug? at the very least should be a `const`